com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java

```diff
@@ -0,0 +1,157 @@
+            // do nothing.
+        }
+
+        SocketListeningConnectorImpl listenConnector = new SocketListeningConnectorImpl(
+                virtualMachineManager());
+        Map<String, Connector.Argument> args = listenConnector.defaultArguments();
+        ((Connector.IntegerArgument) args.get("timeout")).setValue(ACCEPT_TIMEOUT); //$NON-NLS-1$
+        String address = listenConnector.startListening(args);
+
+        String[] cmds = constructLaunchCommand(connectionArgs, address);
+        Process process = Runtime.getRuntime().exec(cmds, envVars, workingDir);
+
+        VirtualMachineImpl vm;
+        try {
+            vm = (VirtualMachineImpl) listenConnector.accept(args);
+        } catch (IOException | IllegalConnectorArgumentsException e) {
+            process.destroy();
+            throw new VMStartException(String.format("VM did not connect within given time: %d ms", ACCEPT_TIMEOUT), process);
+        }
+
+        vm.setLaunchedProcess(process);
+        return vm;
+    }
+
+    private static String[] constructLaunchCommand(Map<String, ? extends Argument> launchingOptions, String address) {
+        String javaHome = launchingOptions.get(HOME).value();
+        String javaOptions = launchingOptions.get(OPTIONS).value();
+        String main = launchingOptions.get(MAIN).value();
+        boolean suspend = Boolean.valueOf(launchingOptions.get(SUSPEND).value());
+        String javaExec = launchingOptions.get(EXEC).value();
+        String slash = System.getProperty("file.separator");
+
+        List<String> cmds = new ArrayList<>();
+        cmds.add(javaHome + slash + "bin" + slash + javaExec);
+        cmds.addAll(parseArguments("-Xdebug -Xnoagent -Djava.compiler=NONE"));
+        cmds.add("-Xrunjdwp:transport=dt_socket,address=" + address + ",server=n,suspend=" + (suspend ? "y" : "n"));
+        cmds.addAll(parseArguments(javaOptions));
+        cmds.addAll(parseArguments(main));
+
+        return cmds.toArray(new String[0]);
+    }
+
+    /**
+     * Parses the given command line into separate arguments that can be passed
+     * to <code>Runtime.getRuntime().exec(cmdArray)</code>.
+     *
+     * @param args command line as a single string.
+     * @return the arguments array.
+     */
+    private static List<String> parseArguments(String cmdStr) {
+        List<String> list = new ArrayList<String>();
+        // The legal arguments are
+        // 1. token starting with something other than quote " and followed by zero or more non-space characters
+        // 2. a quote " followed by whatever, until another quote "
+        Matcher m = Pattern.compile("([^\"]\\S*|\".+?\")\\s*").matcher(cmdStr);
+        while (m.find()) {
+            String arg = m.group(1).replaceAll("^\"|\"$", ""); // Remove surrounding quotes.
+            list.add(arg);
+        }
+        return list;
+    }
+
+    class AdvancedStringArgumentImpl extends StringArgumentImpl implements StringArgument {
```

Comment:
Why do we need to implement this one? Seems the implementation does not do much...

Reply:
The original StringArgumentImpl doesn't expose it's constructor method. Cannot new a StringArgumentImpl in our connector.